Maven Build

fzakaria opened the pull request:
I think there is a lot of benefit to having a standard maven build setup. Maven is definitely one of the de-facto build systems for Java and the integration with tooling (IDEs, testing, reporting) is unmatched.

This is a `WIP` PR to migrate the build process to `maven` while maintaing the current settings and build ordering. It attempts to provide improvements where possible (such as animal-sniffer and checkstyle)
- Added checkstyle as part of build through plugin
- Added toolchain plugin so we can build against JDK1.6 while still using maven (needs 1.7+)
- Added [animal-sniffer](http://www.mojohaus.org/animal-sniffer/) to verify we aren't using any JDK API out of 1.6
- Added test plugin (maven-exec) and disabled surefire since we dont use JUnit
- Correctly modeled the dependencies with appropriate scope
- Added some missing files to gitignore
- Added [Maven wrapper](https://github.com/takari/maven-wrapper) so setup is painless for new developers
- Added README.md that explains use of toolchains.xml and maven wrapper

This PR is with respect to https://github.com/h2database/h2database/issues/276

Commit:
Initial Maven build implementation

 - Added toolchain plugin so we can build against JDK1.6 while still using maven (needs 1.7+)
 - Added animal-sniffer to verify we aren't using any JDK API out of 1.6
 - Added test plugin (maven-exec) and disabled surefire since we dont use JUnit
 - Correctly modeled the dependencies with appropriate scope
 - Added some missing files to gitignore
 - Added Maven wrapper so setup is painless for new developers
 - Added README.md that explains use of toolchains.xml and maven wrapper

fzakaria commented:
I believe these changes to be safe since they've touched no source files and have not touched any build logic- they are an attempt to replicate the current build logic in maven which can be written in tandem.

spyhunter99 commented:
+1

aadrian commented:
+1

grandinj commented:
I'm happy to accept the maven build itself, but not happy to make it the default build process i.e. included in the readme.md

At least not until people have proven they are willing to hang around and maintain the maven build, certainly I don't feel like getting saddled with that responsibility.

So if you move the docs into a separate file, I'd be happy to accept this.

Commit:
Added property resources to JAR

fzakaria commented:
@grandinj 
That sounds good. You want the README moved completely or just reworded to put the maven stuff in a more "temporary" header like "Experimental Builds"

I'll also finish the last few changes to make sure the `TestAll.java` works with args `fast` as the default on the very least.

Side note: Would be cool to then hook this all up with CI/Travis/Coveralls etc..

grandinj commented:
please move the README stuff into a separate file.

I agree, CI integration would be nice

Commit:
Most test passs

- Fixed exec of the TestAll program so that the majority of test passes
- Failures at the moment have to do with missing resources such as web/server/res JSP files

Commit:
reworked documentation

Commit:
Renamed README.md to MAVEN.md

Commit:
Moved to JUnit/Surefire and Added missing resources

- Downloaded H2.jar (off maven) and unziped data.zip and verified we are including all the same resources
- Removed the use of maven-exec-plugin as it was wonky and wrote a bridge class to call TestAll via Junit with "fats" arg.

Commit:
removed checkstyle from pom for now

fzakaria commented:
@grandinj  

I think now its at a good draft for either merge or into a different branch (maven-build).
The test pass rate for **fast** should be identical to the `Build.java` setup (there are some test failures in both).

For now I've created `TestAllJunit.java` (excluded from Build.java) as a bridge for maven's surefire (test) plugin.

I've also moved the `README.md` to `MAVEN.md`

I'd love to get feedback from anyone trying this branch as of now.

Commit:
Added missing test resource optimizations.sql

Commit:
Added instructions for running H2 server

aadrian commented:
> I'd love to get feedback from anyone trying this branch as of now.

Calling `mvnw.cmd clean test` I'm getting:
`Error: Could not find or load main class org.apache.maven.wrapper.MavenWrapperMain`

Commit:
Added missing maven wrapper files

fzakaria commented:
@aadrian  please try now. I forgot two missing files for the maven-wrapper

aadrian commented:
>  I forgot two missing files for the maven-wrapper

Maybe 3.3.9 instead of 3.3.3 ?

Now the wrapper works, but the build process still doesn't, i.e. I'm getting the same error as when I'm calling directly the `mvn` commands. I'm using JDK 8 but this should work since the h2 build tool works with JDK 8 too.

fzakaria commented:
@aadrian (please edit your comment to remove the log lines just to make viewing this PR easier)

You must follow the steps outlined in MAVEN.md to create a `toolchain.xml` file for the following reasons:
- H2 is designed to be run by at minimum the JRE 1.6. Certain classes in the JDK such as `Connection.java` however add new interface methods and exclude us the ability to run from newer JDK (1.7 or 1.8) (or otherwise force us to implement them at which point we invalidate ourselves from JRE 1.6)
- At the moment the current build mechanism has a 'SwitchSource.java' that comments/uncomments blocks of code to accommodate this
- Normally it would be enough to just set `$JAVA_HOME` to JDK 1.6 except that Maven itself needs JDK 1.7+! The solution to this problem is through the use of [toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html)

aadrian commented:
@fzakaria I don't understand. A build with correct 1.6 flags but using a 1.8 should work.
The H2 build tool works with 1.8 so users would expect that the maven based build to work too.

fzakaria commented:
@aadrian They are solving the build in two different ways at the moment. The current build modifies the actual source during compilation, whereas at the moment the maven setup forces you just to specify the proper JDK.

it's a work in progress, but set the `toolchain.xml` to a JDK 1.6 to unblock yourself which anyways better represents the target runtime.
